jobs: fix race which permitted concurrent execution

adityamaru opened the pull request:
The race was that we'd unregister a job from the `adoptedJobs` map while
processing the pause or cancel requests. This is problematic because the
job also unregisters itself in such cases. However, because the job was
already unregistered, the entry which the now canceled job removes may in
fact correspond to a new execution. We remove this hazard by centralizing
the location where the entry is actually removed. Instead, we have the
cancelation/pause processing only cancel the context of the relevant
execution.

Fixes #68593

Release note: None

Release justification: bug fixes and low-risk updates to new functionality

Previously, TraceCollector.StartIter would return a nil object
if we failed to resolve nodeliveness during initialization. This
led to a NPE.

This change now return a TraceCollector instance with the error field
set to the appropriate error, so that the validity check on the iterator
can correctly handle this scenario.

This change also reworks the dump trace on job cancellation test.
Job cancellation semantics under stress are slightly indeterministic in
terms of how many times the execution of OnFailOrCancel is resumed. This makes
it hard to coordinate when to check and how many trace files to expect.

Fixes: #68315

Release note: None

Release justification: low risk, high benefit changes to existing functionality

adityamaru committed:
tracing,job: fix NPE in TraceCollector

Previously, TraceCollector.StartIter would return a nil object
if we failed to resolve nodeliveness during initialization. This
led to a NPE.

This change now return a TraceCollector instance with the error field
set to the appropriate error, so that the validity check on the iterator
can correctly handle this scenario.

This change also reworks the dump trace on job cancellation test.
Job cancellation semantics under stress are slightly undeterministic in
terms of how many times execution of OnFailOrCancel is resumed. This makes
it hard to coordinate when to check and how many trace files to expect.

Fixes: #68315

Release note: None

Release justification: low risk, high benefit changes to existing functionality

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68374)
<!-- Reviewable:end -->

pbardea commented on pkg/jobs/jobs_test.go:
```diff
@@ -1093,42 +1099,47 @@ func TestRegistryLifecycle(t *testing.T) {
 	})
 
 	t.Run("dump traces on cancel", func(t *testing.T) {
+		skip.UnderStress(t, "job cancellation semantics is flaky under stress")
```
If it's flaky due to retries, does it make sense to check that at least a certain number of trace files are there?

pbardea commented on pkg/util/tracing/collector/collector.go:
```diff
@@ -91,7 +91,7 @@ func (t *TraceCollector) StartIter(ctx context.Context, traceID uint64) *Iterato
 	tc := &Iterator{ctx: ctx, traceID: traceID, collector: t}
 	tc.liveNodes, tc.iterErr = nodesFromNodeLiveness(ctx, t.nodeliveness)
 	if tc.iterErr != nil {
```
Should this method be swallowing this error at all? Is it worthwhile to surface?

pbardea reviewed (commented):
Do you know why we're failing to resolve node liveness? From a brief skim, it looks like we'd only fail if the keys weren't there or are otherwise invalid -- is this expected in tests?

adityamaru commented:
>Do you know why we're failing to resolve node liveness? From a brief skim, it looks like we'd only fail if the keys weren't there or are otherwise invalid -- is this expected in tests?

I think it was failing in situations where the call to `maybeDumpTrace` was being invoked after the test had returned. This is actually the same shortcoming that is preventing me from getting the test with cancel to successfully pass under stress. We want to wait for all "resumers" to complete the execution of `maybeDumpTrace` but we don't know how many such resumers are going to be there to be able to wait for all of them.  Checking for `StatusCanceled` at the end does not give us the guarantee that all job resumers have returned from the method. I tried a few approaches one of them is counting how many resumers update the job progress with a new trace ID so that we can wait for those many instances to return before we check the trace files, but there is another issue.

Since some of the resumers get cancelled during `OnFailOrCancel` the order in which https://github.com/adityamaru/cockroach/blob/8a253659275d9e107e073eec4401f5146c647a28/pkg/jobs/jobs_test.go#L298 is set is indeterministic, so we don't know whether to expect a true/false value.

Concretely,
resumer1 <- start OnFailOrCancel
resumer2 <- start OnFailOrCancel
resumer2 <- finish OnFailOrCancel (sets the field to true)
resumer2 <- finish OnFailOrCancel with ctx cancelled (sets the field to false)

ajwerner committed:
jobs: fix race which permitted concurrent execution

The race was that we'd unregister a job from the `adoptedJobs` map while
processing the pause or cancel requests. This is problematic because the
job also unregisters itself in such cases. However, because the job was
already unregistered, the entry which the now canceled job removes may in
fact correspond to a new execution. We remove this hazard by centralizing
the location where the entry is actually removed. Instead, we have the
cancelation/pause processing only cancel the context of the relevant
execution.

Fixes #68593

Release note: None

Release justification: bug fixes and low-risk updates to new functionality

ajwerner reviewed (commented):
> Reviewed 1 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68374)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @dt)

___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 3169 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68374#-Mi6qjOq9xOEzel7Khd9:-Mi6qjOq9xOEzel7KhdA:b-mbaxsk) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e949fa60d8a91eaff0a79a031841362366e2c72/pkg/ccl/changefeedccl/changefeed_test.go#L3169)):*
> ```
> 				unblock()
> 				waitForNoRecord()
> ```
what is this about? Maybe the deal is that we remove the `unblock` channel from `waitForBlocked` and then make it listen for context cancelation. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Thanks for letting me ride on this PR as opposed to making a separate one!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68374#-:-Mi6rbli1kI1RgUdNXej:bsw6nyi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @dt)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

> Reviewed 1 of 10 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68374#-:-Mi7jzPtADAcpLM5YP27:b-jbsf8z)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @dt, and @pbardea)



<!-- Sent from Reviewable.io -->

adityamaru commented:
bors r=pbardea,ajwerner
